### PR TITLE
Route circuits containing CircuitOperations by unrolling them

### LINKED ISSUE #5861: [Routing] Add support for handling CircuitOperations recursively
**Summarize the task**
The current implementation of `RouteCQC` unrolls all CircuitOperations before routing. We want to enhance the transformer by allowing recursive routing of CircuitOperations. 

**Acceptance criteria - when is the task considered done?**
See how this can be done [here](https://docs.google.com/document/d/1df9Onm0Z6ZiU_O7QYRsPtDL1WGTP3330ojEwucI0jZ8/edit#heading=h.dj4v2spnajhy).

Related issues: #5838

### PR DESCRIPTION
**Pull request by Vivek1106-04**
Fixes #5861

RouteCQC can't route a circuit with a CircuitOperation in it unless you pass context.deep=True, and otherwise raises on anything wider than 2 qubits. This unrolls them before routing instead.

I first did this by routing the subcircuits recursively, but that was broken: each one got its own initial mapping and was appended after the outer routing had finished, so it came out on the wrong qubits and in the wrong place. There is one routing path now, over the unrolled circuit, so equivalence follows from the existing algorithm.

CircuitOperations tagged with something in context.tags_to_ignore are left folded and routed as opaque ops, so they still have to fit on 2 qubits.

min_qubit_mapping_threshold (default 0.5) only affects the initial mapping. Unrolling everything just to compute a placement is wasteful when the subcircuits are big, so they get unrolled into the mapper's view one at a time until that fraction of qubits is covered by 2-qubit ops, and the rest stay folded. It never changes the routed circuit. Values outside [0.0, 1.0] raise. Cost of building that view is O(k*m) for k subcircuits of m ops each; measurements are in the comment thread below.

Tests cover unitary equivalence at several thresholds, out of range thresholds, qubit_map / param_resolver / repetitions > 1, tags_to_ignore at top level and nested, subcircuits with only single qubit gates, deep nesting, and directed device graphs.

### CODE REVIEW

**Comment by Vivek1106-04** [comment]
@pavoljuhas ! looks like a flaky test case caused CI run fail

**Comment by Vivek1106-04** [comment]
@tanujkhattar ! Gentle Reminder, Thank you.

**Comment by pavoljuhas** [comment]
Tanuj is too busy with other tasks, reassigning to myself.

**Comment by mhucka** [comment]
@Vivek1106-04 Thank you for this work. I asked some folks at Quantum AI for a quick look. In general, there is positive reception to the idea of including this, with a couple of notes:

- Could you please aim to make it a "contrib" module? (I.e., [cirq-core/cirq/contrib](https://github.com/quantumlib/Cirq/tree/main/cirq-core/cirq/contrib))
- Could you comment about the scaling behavior of this code?
- It looks like this may not produce equivalent circuit behavior. Could you add more test cases with more elaborate circuits, with different kinds of gates, to verify the correctness?

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 1]
```diff
@@ -563,6 +587,87 @@ def _cost(
...
+        for _, _, circuit_op in circuit_ops:
+            inner = circuit_op.circuit.unfreeze(copy=True)
+            inner_routed, inner_init, _ = self.route_circuit(
+                inner,
+                lookahead_radius=lookahead_radius,
+                tag_inserted_swaps=tag_inserted_swaps,
+                initial_mapper=initial_mapper,
+                min_qubit_mapping_threshold=1.0,
+            )
+            routed_circuit.append(circuits.Circuit(inner_routed).transform_qubits(inner_init))
```
Comment (mhucka) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: This alters the execution order. Doesn't that result in a different unitary operation?

**Review by mhucka — CHANGES_REQUESTED** [review]
While adding support for hierarchical or recursive routing of circuits is a valuable goal, based on an initial look at this, it looks like it it may not produce something that is mathematically equivalent.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py (hunk 1 above)
Comment (Vivek1106-04) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: You were right, and it was worse than just reordering.

The moment index was collected into `circuit_ops` and then never used, so every subcircuit got appended after the whole routed circuit. On top of that, each subcircuit was routed with its own fresh initial mapper, so its qubits didn't line up with the outer mapping either.

Small repro on a 4x4 grid at threshold 0.5 — `X` inside a subcircuit, then `H`, then `CNOT`:

```
original:  0: ───X───H───@───     routed:  (1, 1): ───────X───X───
                         │                                   │
           1: ───────────X───              (1, 2): ───H───@───────
```

The mapping was `q0 → (1, 2)`, so the `X` came out both after the `CNOT` and on the wrong qubit.

I don't think this is fixable in place — the outer routing's swaps move the inner qubits, so routing a subcircuit on its own will always drift. Removed the recursive path in e1a81a6d. The threshold now only decides how much of the circuit the initial mapper looks at, and the circuit that gets routed is fully unrolled through the existing `_route`.

**Comment by Vivek1106-04** [comment]
Thanks for taking a look @mhucka!

**Equivalence** — Details are in the inline thread; short version is that subcircuits were routed independently and appended at the end, so both their position and their qubits could be wrong. I've dropped the recursive path. The threshold now only picks how much of the circuit the initial mapper inspects, and routing runs on the fully unrolled circuit, so equivalence comes for free rather than needing an argument.


I also reverted the `LineInitialMapper` change — skipping `CircuitOperation`s there just dropped 2-qubit subcircuits from the interaction graph and made placement worse. That file is out of the PR now.

**Scaling** — the routing algorithm is untouched, so its complexity is unchanged. The only new cost is building the mapping view, `O(k*m)` where `k` is the number of subcircuits actually unrolled, and unrolling stops as soon as the threshold is met. On a 6x6 grid with 20 qubits and subcircuits of 40 moments:

| subcircuits | threshold | ops in mapping view | mapping time |
| --- | --- | --- | --- |
| 2 | 0.1 | 3 | 1.1 ms |
| 2 | 0.99 | 488 | 5.3 ms |
| 8 | 0.1 | 9 | 0.9 ms |
| 8 | 0.99 | 494 | 5.3 ms |
| 32 | 0.1 | 33 | 1.6 ms |
| 32 | 0.99 | 518 | 6.0 ms |

Flat in the number of subcircuits at both ends, since the loop stops at the first one that satisfies the threshold. Happy to fold this into the docstring.

**contrib** — happy to move it, just let me know how you'd like it split. One thing worth deciding either way: the default of `0.5` means a circuit with an untagged 2-qubit `CircuitOperation` now gets unrolled where it used to route as an opaque op with `deep=False`. If that's unwelcome I'll default it to `1.0` so the feature is opt-in.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 2]
```diff
@@ -206,8 +224,19 @@ def route_circuit(
             ValueError: if circuit has operations that act on 3 or more qubits, except measurements.
         """
 
-        # 0. Handle CircuitOperations by unrolling them.
-        if context is not None and context.deep is True:
+        # 0. Handle CircuitOperations by computing the initial mapping from a partially unrolled
+        # view of the circuit and then unrolling them.
+        if initial_mapper is None:
+            initial_mapper = line_initial_mapper.LineInitialMapper(self.device_graph)
+        initial_mapping: dict[cirq.Qid, cirq.Qid] | None = None
```
Comment (mhucka) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:245]: There should be bounds checks somewhere on `min_qubit_mapping_threshold` (e.g., rejecting values <0.0 or >1.0) before the following `if` block.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 3]
```diff
@@ -206,8 +224,19 @@ def route_circuit(
...
+        if self._has_circuit_operations(circuit) and min_qubit_mapping_threshold < 1.0:
+            initial_mapping = initial_mapper.initial_mapping(
+                self._circuit_for_initial_mapping(circuit, min_qubit_mapping_threshold)
+            )
+            circuit = transformer_primitives.unroll_circuit_op(
+                circuit, deep=True, tags_to_check=None
```
Comment (mhucka) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: Calling `unroll_circuit_op(circuit, deep=True, tags_to_check=None)` unconditionally unrolls every `CircuitOperation`, ignoring tags configured in `context.tags_to_ignore`.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 4]
```diff
@@ -206,8 +224,19 @@ def route_circuit(
...
         """
 
-        # 0. Handle CircuitOperations by unrolling them.
-        if context is not None and context.deep is True:
+        # 0. Handle CircuitOperations by computing the initial mapping from a partially unrolled
+        # view of the circuit and then unrolling them.
+        if initial_mapper is None:
+            initial_mapper = line_initial_mapper.LineInitialMapper(self.device_graph)
+        initial_mapping: dict[cirq.Qid, cirq.Qid] | None = None
+        if self._has_circuit_operations(circuit) and min_qubit_mapping_threshold < 1.0:
```
Comment (mhucka) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: If `min_qubit_mapping_threshold = 1.0`, this will skip the `if` branch. If context is `None` (the default), it will fall through without unrolling, hitting `protocols.num_qubits(op) > 2` and raising a `ValueError`. Mathematically, a threshold of 1.0 should mean "unroll until 100% coverage is achieved," not "disable unrolling and crash" :-)

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 5]
```diff
@@ -412,3 +412,172 @@ def test_repr() -> None:
...
+        cirq.CircuitOperation(cirq.Circuit(cirq.CNOT(q[4], q[5]), cirq.CNOT(q[5], q[6])).freeze()),
+    )
+
+    def folded_count(threshold: float) -> int:
+        view = router._circuit_for_initial_mapping(circuit, threshold)
+        return sum(isinstance(op.untagged, cirq.CircuitOperation) for op in view.all_operations())
+
+    assert folded_count(0.2) == 2
+    assert folded_count(0.5) == 1
+    assert folded_count(1.0) == 0
```
Comment (mhucka) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:557]: Calling `router.route_circuit(circuit, min_qubit_mapping_threshold=1.0)` never invokes `_circuit_for_initial_mapping` due to the way it currently checks < 1.0 in `route_circuit`.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py
Comment (mhucka) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:1]: Some additional edge cases to test:

- Out-of-bounds `min_qubit_mapping_threshold` values (<0.0, >1.0)
- `CircuitOperation` with `qubit_map`, `param_resolver`, and `repetitions` > 1
- `CircuitOperation` with `tags_to_ignore` configured in `TransformerContext`
- Subcircuits containing only 1-qubit gates (0 two-qubit coverage)
- Deeply nested `CircuitOperation`s

**Review by mhucka — CHANGES_REQUESTED** [review]

**Comment by mhucka** [comment]
Thank you for your continued efforts on this. An additional general comment:

The PR title may no longer reflect what is happening in the code: this doesn't seem to be doing recursive unrolling. If the goal is simply to allow RouteCQC to accept circuits with CircuitOperations by unrolling them before routing, the PR title, docstrings, and design should be adjusted to reflect that.
